### PR TITLE
fix(core): align Signet prompt with identity stewardship

### PR DESCRIPTION
**Pull request by NicholaiVogel**
## Summary

Refreshes Signet's injected prompt/runtime guidance so agents treat workspace identity files as a durable, self-managed substrate while treating `MEMORY.md` as a system-managed rendered head. Also makes prompt path references workspace-aware instead of assuming `~/.agents`.

## Changes

- rewrote the injected Signet system block in `packages/core/src/markdown.ts` to describe the current layered memory model and hybrid retrieval path more accurately
- made generated prompt/header path references workspace-aware via `$SIGNET_WORKSPACE` or the real workspace path
- added explicit identity stewardship guidance: agents should maintain `AGENTS.md`, `SOUL.md`, `IDENTITY.md`, and `USER.md`, and must not manually edit `MEMORY.md`
- updated setup templates and docs to reinforce the same ownership model
- added `packages/core/src/__tests__/markdown.test.ts` as a regression guard for the prompt-builder contract

## Type

- [ ] `feat` — new user-facing feature (bumps minor)
- [x] `fix` — bug fix
- [ ] `refactor` — restructure without behavior change
- [ ] `chore` — build, deps, config, docs
- [ ] `perf` — performance improvement
- [ ] `test` — test coverage

## Packages affected

- [x] `@signet/core`
- [x] `@signet/daemon`
- [x] `@signet/cli` / dashboard
- [ ] `@signet/sdk`
- [x] `@signet/connector-*`
- [ ] `@signet/web`
- [ ] `predictor`
- [x] Other: docs

## Screenshots

N/A, no UI changes.

## PR Readiness (MANDATORY)

- [x] Spec alignment validated (`INDEX.md` + `dependencies.yaml`)
- [x] Agent scoping verified on all new/changed data queries
- [x] Input/config validation and bounds checks added
- [x] Error handling and fallback paths tested (no silent swallow)
- [x] Security checks applied to admin/mutation endpoints
- [x] Docs updated for API/spec/status changes
- [x] Regression tests added for each bug fix
- [ ] Lint/typecheck/tests pass locally

## Migration Notes (if applicable)

- [ ] Migration is idempotent
- [ ] Daemon Rust parity reviewed or explicitly N/A
- [ ] Rollback / compatibility note included in PR description

## Testing

- [ ] `bun test` passes
- [ ] `bun run typecheck` passes
- [ ] `bun run lint` passes
- [ ] Tested against running daemon
- [ ] N/A

Targeted validation run for this PR:
- `bun test packages/core/src/__tests__/markdown.test.ts packages/core/src/__tests__/identity.test.ts`
- `bun run typecheck` in `packages/core`
- `bun run typecheck` in `packages/connector-base`
- `bun run typecheck` in `packages/connector-opencode`
- `bun run build` in `packages/daemon`

Repo-wide checks are currently blocked by existing unrelated issues:
- `bun run typecheck` fails in `@signet/connector-oh-my-pi` due missing `@types/node`
- `bun run lint` fails on existing repo-wide Biome formatting drift outside this PR

## AI disclosure

- [ ] No AI tools were used in this PR
- [x] AI tools were used (see `Assisted-by` tags in commits)

## Notes

This PR does not add migrations, change retrieval query scoping, or touch admin/mutation endpoint behavior. The readiness items above are marked complete where they are not applicable to the touched surfaces.

### CODE REVIEW

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `9c2aafef`) now and I'll follow up shortly with feedback.

**Comment by PR-Reviewer-Ant** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `9c2aafef`) now and I'll follow up shortly with feedback.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `8459f119`) now and I'll follow up shortly with feedback.

**Comment by PR-Reviewer-Ant** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `8459f119`) now and I'll follow up shortly with feedback.

File: packages/core/src/markdown.ts [hunk 1]
```diff
@@ -15,28 +15,34 @@ export const SIGNET_BLOCK_END = "<!-- SIGNET:END -->";
  * This block ensures agents always know how to use Signet regardless of
  * whether the user's source AGENTS.md was created from template or pre-existed.
  * It provides:
- * - Overview of key files in ~/.agents/
+ * - Overview of key files in the Signet workspace
  * - Dashboard URL
  * - Memory command usage (/remember, /recall)
```
Comment (PR-Reviewer-Ant) [inline packages/core/src/markdown.ts:20]: The new `workspace` interpolation is unsafe in a system-prompt surface. `normalizeWorkspace()` only trims/slashes; it does not escape backticks, brackets, or newlines before embedding `root` into generated markdown that gets synced into harness config files. Because the workspace path is user-configurable, a path containing markdown/control characters can inject or corrupt prompt content in generated `AGENTS.md`/`CLAUDE.md`. This PR expands attack surface compared with the previous hardcoded path. Please escape or sanitize the path before inserting it into prompt text/comments.

File: packages/cli/templates/MEMORY.md.template [hunk 2]
```diff
@@ -3,12 +3,16 @@
 Working memory for {{AGENT_NAME}}. This file is regenerated periodically
 from your memory database.
 
+Do not edit this file manually. It is system-managed by Signet. If you
+need to change durable identity or relationship context, update
+`AGENTS.md`, `SOUL.md`, `IDENTITY.md`, or `USER.md` instead.
+
 > **Related Documents**
```
Comment (PR-Reviewer-Ant) [inline packages/cli/templates/MEMORY.md.template:10]: This template now uses literal `$SIGNET_WORKSPACE/...` strings, but the template itself is static markdown and this PR does not add any substitution step for that variable. That means setup-created files are not actually 'workspace-aware' as the PR description claims; they just contain an unresolved placeholder unless some external harness happens to expand it. The runtime-generated Signet block uses real paths, so the template and injected prompt can now disagree.

**Review by PR-Reviewer-Ant — CHANGES_REQUESTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `9c2aafef`

[Automated Review] Found one prompt-injection risk and one divergence from the PR’s stated workspace-aware behavior.

**Confidence:** Medium [sufficient_diff_evidence, targeted_context_included, missing_cross_module_context] - The prompt builders now interpolate the configured workspace path directly into injected harness prompt text in packages/core/src/markdown.ts without escaping. That is directly visible in the diff and is reachable from the daemon sync path that passes AGENTS_DIR into buildSignetBlock/buildArchitectureDoc. I could not inspect the workspace-path validation layer in this review bundle, so I am assuming POSIX-valid paths may still contain markdown-significant characters or newlines.

File: packages/core/src/markdown.ts [hunk 3]
```diff
@@ -51,7 +57,27 @@ signet recall "user preferences"
...
+---
+
+Treat the files in your Signet workspace as distinct layers of context:
+
+- \`AGENTS.md\` — how you operate
+- \`SOUL.md\` — how you sound, what you value, and your temperament
+- \`IDENTITY.md\` — who you are
+- \`USER.md\` — who the user is and how to relate to them
+- \`MEMORY.md\` — episodic and operational context generated by Signet
+
```
Comment (PR-Reviewer-Ant) [inline packages/core/src/markdown.ts:72]: The new injected guidance hard-codes MCP tool names as `secret_list`, `secret_exec`, `knowledge_expand`, and `knowledge_expand_session`, but the actual exposed Signet tools are namespaced IDs, not these bare names. An agent following this prompt will try to call nonexistent tools and fail on secret/knowledge-graph tasks. Since the PR explicitly claims to make the runtime guidance more accurate, this is a correctness regression in the prompt contract itself.

**Review by PR-Reviewer-Ant — CHANGES_REQUESTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `8459f119`

[Automated Review by pr-reviewer] Found one correctness issue in the new injected prompt guidance.
The PR largely matches its stated goal of updating the Signet prompt/model description and making workspace paths configurable, but the new prompt text now tells agents to call MCP tools by names that do not match the exposed tool IDs. That makes the updated guidance inaccurate in exactly the area this PR is trying to tighten up.

**Confidence:** High [sufficient_diff_evidence, targeted_context_included] - The issue is directly visible in the changed prompt text in packages/core/src/markdown.ts: the injected block now instructs agents to use `secret_list`, `secret_exec`, `knowledge_expand`, and `knowledge_expand_session`, but the exposed Signet MCP tools are namespaced tool IDs rather than those bare names. Because this text is what gets injected into generated harness prompts, the mismatch is directly reachable from the diff without needing runtime reproduction.

**Comment by NicholaiVogel** [comment]
addressed both reviewer callouts:

**MCP tool name mismatch** — updated all four tool references to use the fully-qualified `mcp__signet__` prefix (`mcp__signet__secret_list`, `mcp__signet__secret_exec`, `mcp__signet__knowledge_expand`, `mcp__signet__knowledge_expand_session`). the bare names were inaccurate for Claude Code and other harnesses that namespace by server.

**Workspace path injection** — `normalizeWorkspace` now strips backticks and newlines before interpolating into markdown. backticks are the highest-risk char here since the path appears inside code spans; newlines would break inline structure entirely. added tests covering both edge cases.

all 5 markdown tests pass. commit `86d92195`.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `86d92195`) now and I'll follow up shortly with feedback.

File: packages/cli/templates/AGENTS.md.template [hunk 4]
```diff
@@ -14,22 +14,22 @@ Signet Agent System
 ===
 
 Your identity and memory are managed by Signet, a portable agent identity
-system. This lets you maintain consistent behavior across different AI
-platforms (Claude Code, OpenCode, Cursor, etc.).
+system. It gives you durable identity, working memory, and memory
+retrieval across different AI platforms (Claude Code, OpenCode, Cursor, etc.).
```
Comment (PR-Reviewer-Ant) [inline packages/cli/templates/AGENTS.md.template:19]: This template update is incomplete: later in the same Signet block, the Secrets section still shows `signet secret get OPENAI_API_KEY`. That conflicts with the new runtime guidance in `packages/core/src/markdown.ts` that says agents never see raw secret values and should use the namespaced MCP secret tools instead. New workspaces created from the template will therefore bootstrap agents with a more permissive secret-access pattern than the injected runtime prompt.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `86d92195`

[Automated bot review by pr-reviewer] Found one remaining prompt-guidance inconsistency in the setup template.
The markdown injection and MCP tool-name issues called out earlier are addressed, and the runtime/injected prompt path handling now matches the PR description. The remaining issue is that the updated setup template still leaves agents with a conflicting secret-access instruction.

**Confidence:** Medium [sufficient_diff_evidence, targeted_context_included, missing_cross_module_context] - The diff shows `packages/core/src/markdown.ts` now explicitly telling agents that secrets are exposed only through redacted MCP helpers, while the touched `packages/cli/templates/AGENTS.md.template` still contains a `signet secret get OPENAI_API_KEY` example later in the same Signet block. That creates a concrete mismatch in the guidance new workspaces receive. I did not verify the exact CLI secret command surface in this review bundle, so I am treating the impact as a prompt-policy inconsistency rather than a fully proven runtime exploit.

**Comment by NicholaiVogel** [comment]
addressed the template inconsistency (commit `f5e06501`).

the setup template's Signet block still had `signet secret get OPENAI_API_KEY` / `signet secret list` CLI commands, which contradicts the updated runtime guidance that secrets are MCP-only with redacted injection. updated it to match `buildSignetBlock()` — now shows `mcp__signet__secret_list` and `mcp__signet__secret_exec` consistently.

note: the template block is seed content for new workspaces before first daemon sync; the daemon overwrites it with the generated block on first run, so both paths are now consistent.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `f5e06501`) now and I'll follow up shortly with feedback.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `f5e06501`

[Automated Review by pr-reviewer] I did not find any new bugs, security flaws, or PR-description mismatches in the current diff; the previously flagged prompt-guidance inconsistencies appear addressed, so this looks ready for human review.

**Confidence:** Medium [sufficient_diff_evidence, targeted_context_included, missing_runtime_repro] - The final incremental diff cleanly fixes the last previously identified inconsistency in `packages/cli/templates/AGENTS.md.template` by aligning the template's secret guidance with the updated namespaced MCP tool IDs. I also re-checked the broader shown diff: workspace-aware path rendering is now implemented in `packages/core/src/markdown.ts` and wired through the daemon sync path, with regression tests covering the markdown-sanitization edge cases. I did not run the code or inspect connector-specific runtime outputs beyond the provided diff, so confidence is medium rather than high.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `25719b70`) now and I'll follow up shortly with feedback.

File: packages/daemon/src/daemon.ts [hunk 5]
```diff
@@ -10407,17 +10407,17 @@ async function syncHarnessConfigs() {
 		];
 
 		const existingFiles = files.filter((f) => existsSync(join(AGENTS_DIR, f.name)));
-		const fileList = existingFiles.map((f) => `#   - ~/.agents/${f.name} (${f.desc})`).join("\n");
+		const fileList = existingFiles.map((f) => `#   - ${join(AGENTS_DIR, f.name)} (${f.desc})`).join("\n");
 
 		return `# ${targetName}
 # ============================================================================
-# AUTO-GENERATED from ~/.agents/AGENTS.md by Signet
+# AUTO-GENERATED from ${agentsMdPath} by Signet
```
Comment (PR-Reviewer-Ant) [inline packages/daemon/src/daemon.ts]: The new workspace-aware header now writes `agentsMdPath`/`AGENTS_DIR` directly into generated harness prompt files without any escaping. `buildSignetBlock()` was hardened against backticks/newlines, but this parallel header path is still raw. Because workspace paths are user-controlled, a path containing a newline can break out of the `# ...` header and inject arbitrary prompt text into synced harness files. That is a real prompt-injection surface expansion and means the PR does not fully deliver its claimed path-sanitization hardening.

**Review by PR-Reviewer-Ant — CHANGES_REQUESTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `25719b70`

[Automated Review by pr-reviewer] Found one blocking prompt-injection issue in the new workspace-aware header handling, plus one release-metadata inconsistency in the latest diff.

**Confidence:** High [sufficient_diff_evidence, targeted_context_included] - The blocking issue is directly visible in the changed header generation in packages/daemon/src/daemon.ts: unlike buildSignetBlock(), the newly interpolated workspace/header paths are not normalized before being written into generated harness prompt files. A workspace path containing a newline would break out of the comment header and inject arbitrary prompt text. The release-metadata mismatch is also explicit in the shown diff: package versions were bumped to 0.85.2 while CHANGELOG.md only adds 0.85.0 and 0.85.1 entries.


Unmapped findings (not on changed lines):
- package.json:4 - This bumps the repo version to `0.85.2`, but the same diff only adds CHANGELOG entries for `0.85.0` and `0.85.1`. If this branch is released as-is, the published version will have no matching top-level changelog entry, which breaks release bookkeeping and also introduces unrelated release-train metadata drift beyond what the PR description says it changes.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `7cb2ca8b`) now and I'll follow up shortly with feedback.

**Comment by NicholaiVogel** [comment]
addressed both callouts on `25719b70` (commit `7cb2ca8b`):

**Header path injection (blocking)** — `buildHeader()` in `daemon.ts` was interpolating `AGENTS_DIR` and `agentsMdPath` directly into the generated comment block without sanitizing newlines. added a `safe()` inline helper that strips `\n`/`\r` from all three interpolation points (source path, home path, and per-file paths), matching the same treatment already applied in `normalizeWorkspace()` in `core`.

**CHANGELOG/version mismatch** — this is pre-existing on main, not introduced by this PR. `4ec447c8` (`chore: release 0.85.2`) bumped `package.json` to `0.85.2` on main without a corresponding CHANGELOG entry. this PR doesn't touch either file; the diff the reviewer saw was from the rebase pulling in those main changes. not fixing here — that belongs in a dedicated release chore commit on main.

File: packages/connector-base/src/index.ts [hunk 6]
```diff
@@ -140,11 +140,12 @@ export abstract class BaseConnector {
 	 */
 	protected generateHeader(sourcePath: string, targetName?: string): string {
 		const name = targetName || this.name;
+		const root = dirname(sourcePath);
```
Comment (PR-Reviewer-Ant) [inline packages/connector-base/src/index.ts:145]: This header path interpolation is still unsanitized. `sourcePath` and `dirname(sourcePath)` are derived from the configured workspace path and written directly into generated harness prompt files, so a workspace containing `\n`/`\r` can still break out of the comment header and inject arbitrary prompt text through connector-managed files. The daemon path was fixed for this exact issue in this PR, but the shared connector header builder remains vulnerable; sanitize here as well so all generated headers have the same protection.

**Review by PR-Reviewer-Ant — CHANGES_REQUESTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `7cb2ca8b`

[Automated Review by pr-reviewer] Found one remaining prompt-injection issue in the connector header path handling.

**Confidence:** High [sufficient_diff_evidence, targeted_context_included] - The latest diff fixes newline sanitization in the daemon-generated harness header, but the same workspace-derived header path is still interpolated unsafely in packages/connector-base/src/index.ts. BaseConnector.generateHeader() now derives `root = dirname(sourcePath)` and writes both `sourcePath` and `root` directly into generated prompt files without stripping CR/LF or markdown-delimiting characters. That leaves connector-managed harness files reachable by the same malformed-workspace-path input class the daemon path just fixed.

**Comment by NicholaiVogel** [comment]
fixed in commit `8ba6c199`.

`BaseConnector.generateHeader()` in `packages/connector-base/src/index.ts` had the same pattern — `sourcePath` and `root` interpolated directly into generated comment lines without stripping CR/LF. applied the same `safe()` helper as `daemon.ts`. typecheck passes for `@signet/connector-base`.

**Comment by BusyBee3333** [comment]
Hi @NicholaiVogel - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(core): align Signet prompt with identity stewardship` (commit `8ba6c199`) now and I'll follow up shortly with feedback.

**Review by BusyBee3333 — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `8ba6c199`

Harness output could not be parsed as structured JSON.

```pr-review-json
{}
```

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: gpt-5.4 | commit: `8ba6c199`

[Automated Review by pr-reviewer] I did not find any new bugs, security flaws, or PR-description mismatches in the current diff; this looks ready for human review.

**Confidence:** Medium [sufficient_diff_evidence, targeted_context_included, missing_runtime_repro] - The current head addresses the previously flagged prompt/header injection paths by stripping CR/LF in both daemon and connector-generated headers, and the broader diff consistently updates the injected Signet guidance, templates, and docs to the workspace-aware identity-stewardship model described in the PR. The new regression tests in packages/core/src/__tests__/markdown.test.ts directly cover the prompt-builder contract and markdown sanitization cases. I did not run the code or inspect generated harness outputs end-to-end in a live workspace, so confidence remains medium.
